Optimizer: add test cases for predicate simplifications

ghazalfamilyusa opened the pull request:
Issue Number: ref #40221 

Problem Summary:
This PR just adds tests of various predicate simplification optimizations cases. The purpose of the test is to 
(1) document existing cases. This include (a) removing redundancy for identical simple predicates 
      like =, <>, > and <, >= and <=, (b) un-satisfiable conditions like x=1 and x <> 1 (c) intersection of two in lists.
(2) list cases not working currently and should be addressed in future optimizations. These cases cover 
     the use case in https://github.com/pingcap/tidb/issues/40221) and more like intersection of two "not in" lists.
(3) establish a place holder for predicate simplification is a general optimizer rewrite rule and sub-rules. 

Tests <!-- At least one of them must be included. -->

- [] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

fixdb commented on planner/core/testdata/plan_suite_in.json:
```diff
@@ -1186,5 +1186,22 @@
     ]
+  },
+  {
+    // Test redundant condition in select.
```
We usually add comment in test case function of planner/core/predicate_simplification_test.go, not here.

fixdb commented on planner/core/testdata/plan_suite_out.json:
```diff
@@ -1180,7 +1180,7 @@
       },
       {
-        "SQL": "select f from t use index() where f = 1",
+        "SQL": "select f from t use index() where f = 2",
```
why do we change it to 2?

fixdb commented:
Please sign the CLA and remove the unintentional change, other than that, looks good to me!

ghazalfamilyusa commented:
> Please sign the CLA and remove the unintentional change, other than that, looks good to me!

Removed the unintentional change. For CLA, I signed it yesterday and still showing it is pending. To make sure, I added my ping cap email ahmad.ghazal@pingcap.com to my github account.

fixdb reviewed (approved):
+1
LGTM. We may need help from IT about the CLA.

hawkingrei commented:
@ghazalfamilyusa Please sign the CLA.

ghazalfamilyusa commented:
Hi,
I have done that many time so far. I get that I already signed it and somehow it is not clearing in the PR.

See image of what I keep getting showing that I already accepted.



> On Jan 26, 2023, at 8:36 PM, Weizhen Wang ***@***.***> wrote:
> 
> 
> @ghazalfamilyusa <https://github.com/ghazalfamilyusa> Please sign the CLA.
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/pingcap/tidb/pull/40775#issuecomment-1406010381>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/A3EVQBJQQZAOXIH2HMBHPS3WUNGFNANCNFSM6AAAAAAUG2EF2Y>.
> You are receiving this because you were mentioned.
>

Reminiscent commented on planner/core/predicate_simplification_test.go:
```diff
@@ -0,0 +1,63 @@
+// Copyright 2017 PingCAP, Inc.
```
```suggestion
// Copyright 2023 PingCAP, Inc.
```

Reminiscent commented on planner/core/testdata/plan_suite_in.json:
```diff
@@ -1186,5 +1186,21 @@
+  {
+    "name": "TestRemoveRedundantPredicates",
+    "cases":[
+      "select f from t use index() where f = 1 and f = 1 -- simple redundancy of exact condition",
```
`use index()` hint means we will use the `tableScan` not use the `indexScan`. Why we use this hint in here?

Reminiscent commented:
/merge

hawkingrei commented:
/hold Please resolve the CLA problem.

ghazalfamilyusa commented:
/merge

ghazalfamilyusa commented:
hi @hawkingrei. Anything is holding this merge?

ghazalfamilyusa commented:
Process got complicated when a different email is used for the commit. Will do another clean PR.
